Refactor Client-Example ConsumerConfigs

OwenCorrigan76 opened the pull request:
Refactor the Producer, Consumer and Streams Config files in the Java client by reducing the complexity and size of the class. Standardise the naming scheme of environment variables used to configure the clients / standardise the envVars used to configure the Kafka client properties as outlined here [1].


Example of standardised Environmental Variables:
``` 
KAFKA_<NAME_OF_KAFKA_PROPERTY_DELIMITED_BY_UNDERSCORES>
```

[1]  https://github.com/kyguy/proposals/blob/refactor-client-examples/040-refactor-client-examples.md

OwenCorrigan76 committed:
for refactoring

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
refactored ConsumerConfigs

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
refactored ConsumerConfigs

Signed-off-by: Owen <owencorrigan76@gmail.com>

scholzj reviewed (commented):
I know this is just a draft, so not sure what else do you plan. But I think you should go deeper ... I do not think there is any need to have some special treatment for the TLS certificates or OAuth authentication. I think they can be configured from the `KAFKA_...` env vars as well.

kyguy commented on java/kafka/producer/src/main/java/KafkaProducerConfig.java:
```diff
@@ -210,23 +210,23 @@ public TracingSystem getTracingSystem() {
+                ", tracingSystem='" + tracingSystem + '\'' +
+                '}';
     }
 }
```
Undo changes

OwenCorrigan76 replied:
Fixed

kyguy commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -46,8 +46,7 @@ public static void main(String[] args) {
+                throw new RuntimeException("Error: TRACING_SYSTEM " + tracingSystem + " is not recognized or supported!");
             }
         } else {
             streams = new KafkaStreams(builder.build(), props);
```
undo changes

kyguy commented on java/kafka/deployment-tracing-jaeger.yaml:
```diff
@@ -61,47 +61,6 @@ spec:
-              value: jaeger
----
-apiVersion: apps/v1
-kind: Deployment
```
Undo changes

OwenCorrigan76 replied:
Undone (should be correct now)

kyguy commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -10,175 +10,121 @@
 
         if ((config.getOauthAccessToken() != null)
                 || (config.getOauthTokenEndpointUri() != null && config.getOauthClientId() != null && config.getOauthRefreshToken() != null)
-                || (config.getOauthTokenEndpointUri() != null && config.getOauthClientId() != null && config.getOauthClientSecret() != null))    {
```
undo indentation change

OwenCorrigan76 replied:
Full block of code removed. All security is being handled by the user via the envVars.

kyguy commented on README.md:
```diff
@@ -76,16 +76,13 @@ Producer
+* `KAFKA_GROUP_ID` - specifies the consumer group id for the consumer
+* `STRIMZI_MESSAGE_COUNT` - the number of messages the consumer should receive
+* `STRIMZI_LOG_LEVEL` - logging level  
 * `ADDITIONAL_CONFIG` - additional configuration for a consumer application. Notice, that you can also override any previously set variable by setting this. The form is `key=value` records separated by new line character
```
This can be removed

OwenCorrigan76 replied:
Including the STRIMZI_ prefixed ones?

kyguy commented on README.md:
```diff
@@ -76,16 +76,13 @@ Producer
-* `LOG_LEVEL` - logging level  
+* `KAFKA_BOOTSTRAP_SERVERS` - comma-separated host and port pairs that is a list of Kafka broker addresses. The form of pair is `host:port`, e.g. `my-cluster-kafka-bootstrap:9092` 
+* `STRIMZI_TOPIC` - name of topic which consumer subscribes  
+* `KAFKA_GROUP_ID` - specifies the consumer group id for the consumer
```
Env vars used for Kafka client configuration can be removed.

I would also add a note to the README.md explaining that all Kafka client properties can be configured by adding  an env var with the Kafka property name in a capital letters, delimited by "_", and prefixed with "KAFKA"

OwenCorrigan76 replied:
Done

OwenCorrigan76 commented:
@scholzj When you say "special treatment", do you mean these lines [1] [2]? I kept these in to give the same behaviour as the code before it. Are you saying we can strip this code out completely? If so, are we removing some of the assistance we're offering to the user? So, the user will have to make sure Ssl. Oauth etc. are configured properly from, their end? Thanks

[1] https://github.com/strimzi/client-examples/blob/e40023abb5b7c5d3e1be8a038dca060092e22bde/java/kafka/consumer/src/main/java/KafkaConsumerConfig.java#L90-L99

[2] https://github.com/strimzi/client-examples/blob/e40023abb5b7c5d3e1be8a038dca060092e22bde/java/kafka/consumer/src/main/java/KafkaConsumerConfig.java#L101-L111

scholzj commented:
Yes, that is the code I mean. Also the (de)serializers for example. You can call it an _assistance_. But it also makes it unclear how exactly it is configured. Because one day when they need to write their own app, they will need to dig into the code to understand how to actually do it. I think anything what ends up in the consumer properties file should go directly through the `KAFKA_` environment variables. The only part which should remain are the once which don't end up there ... e.g. number of messages, tracer initiliazation etc.

OwenCorrigan76 committed:
removed Ssl and Oauth configs

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 commented:
@scholzj  @kyguy  I've removed the Ssl and oAuth configs and only left the USER_CONFIGS in. These are the ones that will have a STRIMZI_ prefix. If you guys give me the go ahead that this is the way to go, I'll address the other comments @kyguy left and then move on to the Producer and Streams. Thanks guys.

scholzj commented on README.md:
```diff
@@ -76,16 +76,13 @@ Producer
+* `KAFKA_GROUP_ID` - specifies the consumer group id for the consumer
+* `STRIMZI_MESSAGE_COUNT` - the number of messages the consumer should receive
+* `STRIMZI_LOG_LEVEL` - logging level  
 * `ADDITIONAL_CONFIG` - additional configuration for a consumer application. Notice, that you can also override any previously set variable by setting this. The form is `key=value` records separated by new line character
```
You can remove this as well I guess.

OwenCorrigan76 replied:
@scholzj I've removed ADDITIONAL_CONFIG as it's no longer part of ConsumerConfig. Do you suggest removing the STRIMZI_ prefixed envVar examples too and just leave in the explanation at the beginning of the ##Configuration paragraph?

OwenCorrigan76 replied:
.yaml files updated with correct prefixes.

scholzj commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,85 @@
+                ", enableAutoCommit='" + enableAutoCommit + '\'' +
+                ", messageCount=" + messageCount +
+                ", tracingSystem='" + tracingSystem + '\'' +
+                kafkaConfigOptionsToString() +
```
Is this needed? The Kafka client prints its config at startup. If you want to keep this, you should check the formatting to fit into the other fields.

OwenCorrigan76 replied:
Not needed now as we have ` kafkaConfigOptionsToString()`. Removed

scholzj commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,85 @@
+            .entrySet()
+            .stream()
+            .filter(map -> map.getKey().startsWith(KAFKA_PREFIX))
+            .collect(Collectors.toMap(map -> convertEnvVarToPropertyKey(map.getKey()), map -> map.getValue()));
```
1) Should you use Properties directly here?
2) I also wonder if this should be static. In general, for things such as testing etc., it would be better to handle this in the constructor or in the `fromEnv` method. You might nto test it today, but maybe you can add some tests for this in the future.

OwenCorrigan76 replied:
Discuss today with @kyguy

scholzj commented on java/kafka/producer/src/main/java/KafkaProducerConfig.java:
```diff
@@ -210,23 +210,23 @@ public TracingSystem getTracingSystem() {
-            ", additionalConfig='" + additionalConfig + '\'' +
-            ", tracingSystem='" + tracingSystem + '\'' +
-            '}';
+                "bootstrapServers='" + bootstrapServers + '\'' +
+                ", topic='" + topic + '\'' +
+                ", delay=" + delay +
+                ", messageCount=" + messageCount +
+                ", message='" + message + '\'' +
+                ", acks='" + acks + '\'' +
+                ", headers='" + headers + '\'' +
+                ", sslTruststoreCertificates='" + sslTruststoreCertificates + '\'' +
+                ", sslKeystoreKey='" + sslKeystoreKey + '\'' +
+                ", sslKeystoreCertificateChain='" + sslKeystoreCertificateChain + '\'' +
+                ", oauthClientId='" + oauthClientId + '\'' +
+                ", oauthClientSecret='" + oauthClientSecret + '\'' +
+                ", oauthAccessToken='" + oauthAccessToken + '\'' +
+                ", oauthRefreshToken='" + oauthRefreshToken + '\'' +
+                ", oauthTokenEndpointUri='" + oauthTokenEndpointUri + '\'' +
+                ", additionalConfig='" + additionalConfig + '\'' +
+                ", tracingSystem='" + tracingSystem + '\'' +
```
Maybe this should not be part of this PR? Ideally, you should try to keep the PRs focused on a single thing.

OwenCorrigan76 replied:
Will remove

scholzj commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -46,8 +46,7 @@ public static void main(String[] args) {
             } else {
-                log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
-                streams = new KafkaStreams(builder.build(), props);
+                throw new RuntimeException("Error: TRACING_SYSTEM " + tracingSystem + " is not recognized or supported!");
```
Same as above ... Maybe this should not be part of this PR? Ideally, you should try to keep the PRs focused on a single thing.

OwenCorrigan76 replied:
Will remove

kyguy replied:
Revert to original

scholzj commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,85 @@
     private final String topic;
-    private final String groupId;
-    private final String autoOffsetReset = "earliest";
     private final String enableAutoCommit = "false";
```
This seems to be a weird left-over. You do not set it, but you keep it in the code. If it is needed, maybe you should check if it is enabled or not through the Kafka properties and set it from there? (https://kafka.apache.org/documentation/#consumerconfigs_enable.auto.commit)

OwenCorrigan76 replied:
autoOffsetReset removed. Discuss enableAutoCommit with @kyguy

scholzj replied:
I think it is a bit more complicated then just removing it. The auto-commit is part of Kafka. But what behaviour do you expect when it is disabled? How does the client do the commits etc.? That is something you need to consider.

scholzj commented on README.md:
```diff
@@ -76,16 +76,13 @@ Producer
-* `USER_CRT` - the user's certificate
-* `USER_KEY` - the user's private key
-* `LOG_LEVEL` - logging level  
+* `KAFKA_BOOTSTRAP_SERVERS` - comma-separated host and port pairs that is a list of Kafka broker addresses. The form of pair is `host:port`, e.g. `my-cluster-kafka-bootstrap:9092` 
+* `STRIMZI_TOPIC` - name of topic which consumer subscribes  
+* `KAFKA_GROUP_ID` - specifies the consumer group id for the consumer
+* `STRIMZI_MESSAGE_COUNT` - the number of messages the consumer should receive
+* `STRIMZI_LOG_LEVEL` - logging level  
```
You should expans this a bit. List the Strimzi specific options and elaborate that any Kafka otpions can be used as Env var by:
* Preficing it with `KAFKA_` and making it uppercase and using `_` instead of `.`.

OwenCorrigan76 replied:
Added:
Below are listed and described environmental variables that are NOT Kafka-specific but must be used. These environmental variables should be prefixed
with uppercase STRIMZI_  using _ instead of . (eg. STRIMZI_TOPIC). Any Kafka configuration option can be used as an environmental variable and should be prefixed with uppercase KAFKA_ using _ instead of . (eg. KAFKA_BOOTSTRAP_SERVERS).

scholzj reviewed (commented):
Some more comments, but it starts looking better. You should also update all the YAML example files to make it work with the new code.

OwenCorrigan76 committed:
Indentation in jaeger.yaml

Signed-off-by: Owen <owencorrigan76@gmail.com>

kyguy commented on java/kafka/consumer/src/main/java/KafkaConsumerExample.java:
```diff
@@ -41,7 +43,7 @@ public static void main(String[] args) {
         }
 
-        boolean commit = !Boolean.parseBoolean(config.getEnableAutoCommit());
+       boolean commit = !Boolean.parseBoolean (props.getProperty(ENABLE_AUTO_COMMIT_CONFIG));
```
```suggestion
       boolean commit = !Boolean.parseBoolean(props.getProperty(ENABLE_AUTO_COMMIT_CONFIG));
```

kyguy commented on java/kafka/deployment-ssl.yaml:
```diff
@@ -110,18 +110,18 @@ spec:
+          - name: KAFKA_GROUP_ID
             value: java-kafka-consumer
-          - name: LOG_LEVEL
+          - name: STRIMZI_ LOG_LEVEL
```
```suggestion
          - name: STRIMZI_LOG_LEVEL
```

kyguy commented on java/kafka/deployment-tracing-jaeger.yaml:
```diff
@@ -3,7 +3,7 @@ kind: KafkaTopic
   name: my-topic
   labels:
-      strimzi.io/cluster: my-cluster
+    strimzi.io/cluster: my-cluster
```
extra two spaces removed

kyguy commented on java/kafka/deployment-tracing-jaeger.yaml:
```diff
@@ -117,26 +76,26 @@ spec:
+          - name: JAEGER_SAMPLER_PARAM
+            value: "1"
+          - name: TRACING_SYSTEM
+            value: jaeger
```
add two space indentation back

kyguy commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -47,8 +47,7 @@ public static void main(String[] args) {
                 log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
-                streams = new KafkaStreams(builder.build(), props);
-            }
+                streams = new KafkaStreams(builder.build(), props);   }
```
Rogue formatting change, can be undone

kyguy commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -47,8 +47,7 @@ public static void main(String[] args) {
             } else {
                 log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
-                streams = new KafkaStreams(builder.build(), props);
-            }
```
Rogue formatting change, can be undone

kyguy reviewed (commented):
Just some formatting nits

scholzj commented on README.md:
```diff
@@ -57,7 +57,9 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
 Although this Hello World is simple example it is fully configurable.
-Below are listed and described environmental variables.
+Below are listed and described environmental variables that are NOT Kafka-specific but must be used. 
+These environmental variables should be prefixed with `STRIMZI_` using `_` instead of `.` (eg. STRIMZI_TOPIC). 
```
These are just random environment variables. The `_` versus `.` makes no sense here. It makes sense only for the `KAFKA_` env vars which map to actual Kafka configuration properties.

OwenCorrigan76 replied:
Removed and refactored to fit the uniformity of the rest of the README.

scholzj commented on README.md:
```diff
@@ -57,7 +57,9 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
-Below are listed and described environmental variables.
+Below are listed and described environmental variables that are NOT Kafka-specific but must be used. 
+These environmental variables should be prefixed with `STRIMZI_` using `_` instead of `.` (eg. STRIMZI_TOPIC). 
+Any Kafka configuration option can be given as an environmental variable and should be prefixed with `KAFKA_` using `_` instead of `.` (eg. bootstrap.servers -> KAFKA_BOOTSTRAP_SERVERS).
```
```suggestion
Any Kafka configuration option can be given as an environmental variable and should be prefixed with `KAFKA_` using `_` instead of `.` (eg. `bootstrap.servers` -> `KAFKA_BOOTSTRAP_SERVERS`).
```

OwenCorrigan76 replied:
Added back in

scholzj commented on README.md:
```diff
@@ -57,7 +57,9 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
 
 Although this Hello World is simple example it is fully configurable.
-Below are listed and described environmental variables.
+Below are listed and described environmental variables that are NOT Kafka-specific but must be used. 
+These environmental variables should be prefixed with `STRIMZI_` using `_` instead of `.` (eg. STRIMZI_TOPIC). 
+Any Kafka configuration option can be given as an environmental variable and should be prefixed with `KAFKA_` using `_` instead of `.` (eg. bootstrap.servers -> KAFKA_BOOTSTRAP_SERVERS).
```
So, does this currently apply to consumers only? It is fine to do it step by step. But you should make it clear here in the README. Right now it sounds like this applies for producers and Streams as well.

OwenCorrigan76 replied:
Refactored to specify Consumer only (until later updates)

scholzj commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,56 @@
-    public String getBootstrapServers() {
-        return bootstrapServers;
+    public static String convertEnvVarToPropertyKey(String envVar) {
+        System.out.println("ENV_VAR is " + envVar);
```
This should use logger if you want to log this (doesn't seem needed to me, but why not) or be removed if it is some leftover from debugging.

OwenCorrigan76 replied:
Removed altogether

scholzj reviewed (commented):
Left few nits mostl around formatting etc. But looks good overall.

OwenCorrigan76 committed:
Addressing Jakub's Comments and changing Repo name

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 commented:
@scholzj @kyguy Thanks for the reviews / feedback. I've made the latest changes as commented by @scholzj. I will now start to refactor the ProducerConfigs. As I am doing so, I think it prudent to refactor the name of the PR to Refactor Client-Example Configs.

ppatierno commented on README.md:
```diff
@@ -57,7 +57,10 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
 
 Although this Hello World is simple example it is fully configurable.
-Below are listed and described environmental variables.
+Below are listed and described environmental variables. 
```
I think we use "environment variable" everywhere in our docs. Can we change this way everywhere you are using "environmental"?

OwenCorrigan76 replied:
Done

ppatierno commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,55 @@
+                .entrySet()
+                .stream()
+                .filter(map -> map.getKey().startsWith(KAFKA_PREFIX))
+                .collect(Collectors.toMap(map -> convertEnvVarToPropertyKey(map.getKey()), map -> map.getValue())));
```
is `map` appropriate as variable here? You are getting a `Map.Entry` instance right? I would be more for naming it `entry` or `mapEntry` but not just `map` which makes reader thinking about the whole map itself.

OwenCorrigan76 replied:
Refactored to use mapEntry

ppatierno commented on java/kafka/consumer/src/main/java/KafkaConsumerExample.java:
```diff
@@ -41,7 +43,7 @@ public static void main(String[] args) {
         }
 
-        boolean commit = !Boolean.parseBoolean(config.getEnableAutoCommit());
+       boolean commit = !Boolean.parseBoolean(props.getProperty(ENABLE_AUTO_COMMIT_CONFIG));
```
wrong indentation, missing space?

OwenCorrigan76 replied:
Fixed

scholzj commented:
> @scholzj @kyguy Thanks for the reviews / feedback. I've made the latest changes as commented by @scholzj. I will now start to refactor the ProducerConfigs. As I am doing so, I think it prudent to refactor the name of the PR to Refactor Client-Example Configs.

Too late I guess. But we should have first approved and merged the consumer config you had. This just makes it harder to review and fix things.

kyguy commented on java/kafka/producer/src/main/java/KafkaProducerConfig.java:
```diff
@@ -3,230 +3,85 @@
-        return props;
+        String topic = System.getenv("STRIMZI_TOPIC");
+        Long messageCount = System.getenv("STRIMZI_MESSAGE_COUNT") == null ? DEFAULT_MESSAGES_COUNT : Long.parseLong(System.getenv("STRIMZI_MESSAGE_COUNT"));
+        int delay = System.getenv("STRIMZI_DELAY") == null ? DEFAULT_DELAY : Integer.parseInt(System.getenv("STRIMZI_DELAY"));
```
```suggestion
        int delay = Integer.parseInt(System.getenv("STRIMZI_DELAY_MS"));
```

kyguy commented on java/kafka/producer/src/main/java/KafkaProducerConfig.java:
```diff
@@ -3,230 +3,85 @@
     private static final long DEFAULT_MESSAGES_COUNT = 10;
     private static final String DEFAULT_MESSAGE = "Hello world";
-    private final String bootstrapServers;
+    private static final int DEFAULT_DELAY = 1;
```
Can be removed

kyguy commented on java/kafka/producer/src/main/java/KafkaProducerExample.java:
```diff
@@ -19,6 +19,7 @@
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicLong;
 
+
```
remove

kyguy commented on java/kafka/producer/src/main/java/KafkaProducerExample.java:
```diff
@@ -57,8 +58,11 @@ public static void main(String[] args) throws InterruptedException {
-        boolean transactionalProducer = config.getAdditionalConfig().contains("transactional.id");
+      //  boolean transactionalProducer = !Boolean.parseBoolean(props.getProperty(TRANSACTIONAL_ID_CONFIG));
+        boolean transactionalProducer = config.getProperties().contains("transactional.id");
+      //  boolean transactionalProducer = config.getAdditionalConfig().contains("transactional.id");
```
Don't forget to clean up

OwenCorrigan76 replied:
Reverted to before refactoring ConsumerConfig

kyguy commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -46,13 +46,12 @@ public static void main(String[] args) {
                 streams = new KafkaStreams(builder.build(), props, supplier);
             } else {
-                log.error("Error: TRACING_SYSTEM {} is not recognized or supported!", config.getTracingSystem());
-                streams = new KafkaStreams(builder.build(), props);
```
Add back in, revert to original

OwenCorrigan76 replied:
Done

OwenCorrigan76 committed:
Addressing Comments

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Addressing Kyle's Comments

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Addressing Paolo's Comments

Signed-off-by: Owen <owencorrigan76@gmail.com>

kyguy commented on java/kafka/streams/src/main/java/KafkaStreamsExample.java:
```diff
@@ -54,4 +54,4 @@ public static void main(String[] args) {
 
         streams.start();
     }
-}
```
Undo

ppatierno commented:
@OwenCorrigan76 can we move out of the draft state?

ppatierno commented:
Agree with @scholzj ... but it seems not to be late yet. There are no changes on the producer side yet :-)

scholzj commented on java/kafka/consumer/src/main/java/KafkaConsumerConfig.java:
```diff
@@ -3,228 +3,56 @@
-
-    public String getBootstrapServers() {
-        return bootstrapServers;
+    public static String convertEnvVarToPropertyKey(String envVar) {
```
Can be private I guess?

OwenCorrigan76 replied:
Refactored

scholzj commented on README.md:
```diff
@@ -57,7 +57,10 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
 Although this Hello World is simple example it is fully configurable.
-Below are listed and described environmental variables.
+Below are listed and described environment variables. 
+Regarding Consumer Only (until later updates): below are listed and described environmental variables that are NOT Kafka-specific but must be used. 
```
Only with lowercase `o` I guess? Also, is `must be used` the right way?

OwenCorrigan76 replied:
Refactored

scholzj commented on README.md:
```diff
@@ -57,7 +57,10 @@ To run the OAuth example, you will need to have your Kafka cluster configured wi
 ## Configuration
```
Maybe you can do something like this (just the `Configuration` section):

```markdown
## Configuration

Below are listed and described the available environment variables that can be used for configuration. 

### Producer  

* `BOOTSTRAP_SERVERS` - comma-separated host and port pairs that is a list of Kafka broker addresses. The form of pair is `host:port`, e.g. `my-cluster-kafka-bootstrap:9092` 
* `TOPIC` - the topic the producer will send to  
* `DELAY_MS` - the delay, in ms, between messages  
* `MESSAGE_COUNT` - the number of messages the producer should send  
* `CA_CRT` - the certificate of the CA which signed the brokers' TLS certificates, for adding to the client's trust store
* `USER_CRT` - the user's certificate
* `USER_KEY` - the user's private key
* `LOG_LEVEL` - logging level  
* `PRODUCER_ACKS` - acknowledgement level
* `HEADERS` - custom headers list separated by commas of `key1=value1, key2=value2`
* `BLOCKING_PRODUCER` - if it's set, the producer will block another message until ack will be received
* `MESSAGES_PER_TRANSACTION` - how many messages will be part of one transaction. Transaction config could be set via `ADDITIONAL_CONFIG` variable. Default is 10.
* `ADDITIONAL_CONFIG` - additional configuration for a producer application. Notice, that you can also override any previously set variable by setting this. The form is `key=value` records separated by new line character
* `TRACING_SYSTEM` - if it's set to `jaeger` or `opentelemetry`, this will enable tracing. 

### Consumer  

* `STRIMZI_TOPIC` - name of topic to which consumer subscribes  
* `STRIMZI_MESSAGE_COUNT` - the number of messages the consumer should receive
* `STRIMZI_LOG_LEVEL` - logging level  
* `STRIMZI_TRACING_SYSTEM` - if it's set to `jaeger` or `opentelemetry`, this will enable tracing.

Additionally, any Kafka Consumer API configuration option can be passed as an environmental variable.
It should be prefixed with `KAFKA_` and use `_` instead of `.`.
For example environment variable `KAFKA_BOOTSTRAP_SERVERS` will be used as the `bootstrap.servers` configuration option in the Kafka Consumer API.

### Streams  

* `BOOTSTRAP_SERVERS` - comma-separated host and port pairs that is a list of Kafka broker addresses. The form of pair is `host:port`, e.g. `my-cluster-kafka-bootstrap:9092`
* `APPLICATION_ID` - The Kafka Streams application ID
* `SOURCE_TOPIC` - name of topic which will be used as the source of messages
* `TARGET_TOPIC` - name of topic where the transformed images are sent
* `COMMIT_INTERVAL_MS` - the interval for the Kafka Streams consumer part committing the offsets
* `CA_CRT` - the certificate of the CA which signed the brokers' TLS certificates, for adding to the client's trust store
* `USER_CRT` - the user's certificate
* `USER_KEY` - the user's private key
* `LOG_LEVEL` - logging level
* `ADDITIONAL_CONFIG` - additional configuration for a streams application. Notice, that you can also override any previously set variable by setting this. The form is `key=value` records separated by new line character.
* `TRACING_SYSTEM` - if it's set to `jaeger` or `opentelemetry`, this will enable tracing.

### Tracing

The examples support tracing using the [OpenTracing Apache Kafka Instrumentation](https://github.com/opentracing-contrib/java-kafka-client), 
[OpenTelemetry Java Instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and the [Jaeger project](https://www.jaegertracing.io/).
To enable tracing, configure the Jaeger Tracer using [environment variables](https://github.com/jaegertracing/jaeger-client-java/tree/master/jaeger-core#configuration-via-environment).
To run Jaeger Tracing, you can also use the provided example in [`deployment-tracing-jaeger.yaml`](./java/kafka/deployment-tracing-jaeger.yaml).

To run OpenTelemetry Tracing, you can also use the provided example in [`deployment-tracing-opentelemetry.yaml`](./java/kafka/deployment-tracing-opentelemetry.yaml).

Jaeger / OpenTracing tracing is supported only in consumers / producers because OpenTracing support for Kafka Streams API is not compatible with the latest Kafka versions.
```

OwenCorrigan76 replied:
Makes sense. Refactored to use this suggestion.

OwenCorrigan76 committed:
Addressing Jakub and Paolo's Comments

Signed-off-by: Owen <owencorrigan76@gmail.com>

OwenCorrigan76 commented:
@scholzj @ppatierno I had done some work on ProducerConfigs but I reverted everything back to before I changed ProducerConfigs. I guess I'll concentrate on Consumer first. Get it merged and then open a new PR for Producer. When that is merged, I'll open a separate PR for Streams. This means that any of Kyle's comments regarding ProducerConfigs above can be ignored. I know it's messy but i originally thought you guys would prefer Producer / Consumer and Streams done in the same PR (my mistake).

OwenCorrigan76 committed:
Further Addressing Jakub's Comments

Signed-off-by: Owen <owencorrigan76@gmail.com>

scholzj reviewed (approved):
Thanks for the PR!
